File: Fore-C_v2/codes/custom_functions/fun_pixels_to_management_zones.R
library(tidyverse)

agg_to_manage_zones_forecasts <- function(forecast, zone_polygon_with_id, diseaseRegion, fileName){
  x <- merge(
    forecast
    , zone_polygon_with_id
    , by.x = "ID"
    , by.y = "PixelID"
    )
  
  x_agg <- x %>%
    group_by(
      PolygonID
      , Region
      , Date
      , type
      ) %>%
    summarise(
      Lwr = quantile(Lwr, 0.90)
      , value = quantile(value, 0.90)
      , Upr = quantile(Upr, 0.90)
      )
  
  if(diseaseRegion == "ga_gbr"){
    nostress_threshold <- 5
    watch_threshold <- 10
    warning_threshold <- 15
    alert1_threshold <- 25
  } else if(diseaseRegion == "ga_pac"){
    nostress_threshold <- 5
    watch_threshold <- 10
    warning_threshold <- 15
    alert1_threshold <- 25
  } else if(diseaseRegion == "ws_gbr"){
    nostress_threshold <- 1
    watch_threshold <- 5
    warning_threshold <- 10
    alert1_threshold <- 20
  } else if(diseaseRegion == "ws_pac"){
    nostress_threshold <- 1
    watch_threshold <- 5
    warning_threshold <- 10
    alert1_threshold <- 15
  } 
  
  x_agg$drisk <- NA
  x_agg$drisk[x_agg$value >= 0 & x_agg$value <= nostress_threshold] <- 0
  x_agg$drisk[x_agg$value > nostress_threshold & x_agg$value <= watch_threshold] <- 1
  x_agg$drisk[x_agg$value > watch_threshold & x_agg$value <= warning_threshold] <- 2
  x_agg$drisk[x_agg$value > warning_threshold & x_agg$value <= alert1_threshold] <- 3
  x_agg$drisk[x_agg$value > alert1_threshold] <- 4
  
  as.data.frame(x_agg)
}

agg_to_manage_zones_scenarios <- function(predictions, zone_polygon_with_id, fileName){
  aggregated_forecast <- merge(predictions,
                               zone_polygon_with_id,
                               by.x = "ID",
                               by.y = "PixelID")
  
  aggregated_forecast_to_management_zone <- aggregated_forecast %>%
    group_by(PolygonID, Region, Response, Response_level) %>%
    summarise_at(vars(value, estimate, LwrEstimate, UprEstimate, disease_risk_change), median) %>%
    mutate(ID = PolygonID) 
  
  aggregated_forecast_to_management_zone$PolygonID <- NULL
  
  as.data.frame(aggregated_forecast_to_management_zone)
  
}

File: Fore-C_v2/codes/custom_functions/fun_pixels_to_management_zones.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 31 21:29:35 2022

@author: jamie
"""

def agg_to_manage_zones_forecasts(forecast, management_df, dz):
    # merge datasets
    df = forecast.merge(management_df, left_on = 'ID', right_on = 'PixelID')
    # drop columns
    df = df.drop(['PixelID', 'Latitude', 'Longitude', 'drisk'], axis = 1) 
    # get quantile values
    df2 = df.groupby(['PolygonID', 'Region', 'Date', 'type']).quantile(0.90).reset_index()
    # set thresholds by disease-region
    if dz == 'ga':
        nostress_threshold = 5
        watch_threshold = 10
        warning_threshold = 15
        alert1_threshold = 25
    else:
        nostress_threshold = 1
        watch_threshold = 5
        warning_threshold = 10
    
    # assign risk level
    df2['drisk'] = ''
    # no stress
    nostress_ind = df2.index[(df2['value'] >= 0) & (df2['value'] <= nostress_threshold)].tolist()
    df2.loc[nostress_ind, ('drisk')] = 0
    # watch
    watch_ind = df2.index[(df2['value'] > nostress_threshold) & (df2['value'] <= watch_threshold)].tolist()
    df2.loc[watch_ind, ('drisk')] = 1
    # warning
    warn_ind = df2.index[(df2['value'] > watch_threshold) & (df2['value'] <= warning_threshold)].tolist()
    df2.loc[warn_ind, ('drisk')] = 2
    if dz == 'ga':
        # alert level 1
        alert1_ind = df2.index[(df2['value'] > warning_threshold) & (df2['value'] <= alert1_threshold)].tolist()
        # alert level 2
        alert2_ind = df2.index[(df2['value'] > alert1_threshold)].tolist()
    else:
        # alert level 1
        alert1_ind_gbr = df2.index[(df2['Region'] == 'gbr') & (df2['value'] > warning_threshold) & (df2['value'] <= 20)].tolist()
        alert1_ind_pac = df2.index[(df2['Region'] != 'gbr') & (df2['value'] > warning_threshold) & (df2['value'] <= 15)].tolist()
        alert1_ind = alert1_ind_gbr + alert1_ind_pac
        # alert level 2
        alert2_ind_gbr = df2.index[(df2['Region'] == 'gbr') & (df2['value'] > 20)].tolist()
        alert2_ind_pac = df2.index[(df2['Region'] != 'gbr') & (df2['value'] > 15)].tolist()
        alert2_ind = alert2_ind_gbr + alert2_ind_pac
    # alert level 1
    df2.loc[alert1_ind, ('drisk')] = 3
    # alert level 2
    df2.loc[alert2_ind, ('drisk')] = 4
    # return dataframe
    return df2


def agg_to_manage_zones_scenarios(scenario, management_df):
    # merge data
    df = scenario.merge(management_df, left_on = 'ID', right_on = 'PixelID')
    # get median values by group
    df = df.groupby(['PolygonID', 'Region', 'Response', 'Response_level']).median().reset_index()
    # format columns
    df['ID'] = df['PolygonID']
    df = df.drop(['PolygonID', 'Latitude', 'Longitude', 'PixelID'], axis = 1)
    # return data
    return df
    
